packages/pwa-kit-react-sdk/src/webpack/plugins.js

```diff
@@ -91,3 +95,104 @@ export class BuildMarkerPlugin {
+ */
+export class PwaKitConfigPlugin {
+    /**
+     * @constructor
+     * @param {String} [options.path] - The relative path of the config file
+     * @param {Object[]} [options.customSchemas] - The object that include
+     * custom property schemas, you want to use this when you extend the schema.
+     * @param {String} [options.customSchemas.key] - The key of the property.
+     * @param {Boolean} [options.customSchemas.required] - Required.
+     * @param {Object} [options.customSchemas.schema] - The JSON schema for the
+     * custom property.
+     */
+    constructor(options = {}) {
+        this.path = options.path || './pwa-kit.config.json'
+        this.customSchemas = options.customSchemas || []
+    }
+
+    /**
+     * Called by webpack when this plugin is attached.
+     *
+     * @param compiler {Object} the webpack compiler
+     */
+    apply(compiler) {
+        compiler.hooks.compilation.tap('PwaKitConfigPlugin', (compilation) => {
+            try {
+                const config = this.getConfig(compiler)
+                this.validate(config)
+            } catch (e) {
+                compilation.errors.push(e)
+            }
+        })
+    }
+
+    /**
+     * Find the config file from project and return the value.
+     *
+     * @param compiler {Object} the webpack compiler
+     * @returns {object} - The configuration object
+     */
+    getConfig(compiler) {
+        const {fileSystem} = compiler.inputFileSystem
+        const absolutePath = path.resolve(this.path)
+        let configFile
+        try {
+            configFile = fileSystem.readFileSync(absolutePath)
+        } catch (e) {
+            throw new Error(`Missing PWA Kit config file, the file is required: ${absolutePath}`)
+        }
+        return JSON.parse(configFile)
+    }
+
+    /**
+     * Validate configurations based on pwa-kit-react-sdk schema
+     * and custom schema. Errors will be thrown when validation fails.
+     *
+     * @param config {Object} - the configuration object
+     */
+    validate(config) {
+        const ajv = new Ajv()
+        AjvMergePlugin(ajv)
+        ajv.addSchema(schema)
+        let valid
+
+        if (this.customSchemas.length) {
+            const required = [...schema.required]
+            const properties = this.customSchemas.reduce((prev, curr) => {
+                if (curr.required) {
+                    required.push(curr.key)
+                }
+                return {...prev, [curr['key']]: curr.schema}
+            }, {})
+            valid = ajv.validate(
+                {
+                    $merge: {
+                        source: {$ref: schema.$id},
+                        with: {
+                            required,
+                            properties
+                        }
+                    }
+                },
+                config
+            )
+        } else {
+            valid = ajv.validate(schema, config)
+        }
```

Comment:
I think I understand what you are doing here. If there is a custom schema you update the app config schema from the sdk with any required properties and other properties. But why do you have to treat the required ones and the rest in a special way.  Maybe you can help me understand it a little more.

Reply:
Let's use an example here, let's say a user extends the schema and added a new `commerceApi` property.

The schema for the new property according to JSON schema specification looks like:
```
{
            $id: 'commerceApi',
            title: 'Commerceapi',
            type: 'object',
            required: ['clientId', 'organizationId', 'shortCode', 'siteId'],
            properties: {
                clientId: {
                    $id: 'commerceApi/clientId',
                    title: 'Clientid',
                    type: 'string',
                    default: '',
                    examples: ['11111111-1111-1111-1111-111111111111'],
                    pattern: '^.*$'
                },
                organizationId: {
                    $id: 'commerceApi/organizationId',
                    title: 'Organizationid',
                    type: 'string',
                    default: '',
                    examples: ['f_ecom_zzrf_001'],
                    pattern: '^.*$'
                },
                shortCode: {
                    $id: 'commerceApi/shortCode',
                    title: 'Shortcode',
                    type: 'string',
                    default: '',
                    examples: ['9o7m175y'],
                    pattern: '^.*$'
                },
                siteId: {
                    $id: 'commerceApi/siteId',
                    title: 'Siteid',
                    type: 'string',
                    default: '',
                    examples: ['RefArchGlobal'],
                    pattern: '^.*$'
                }
            }
```

If you look closely at `required`, you might imagine the `required` field is a boolean value, but it's not, JSON schema specification use `required` to declare the list of required descendants. So, there is a limitation with JSON schema where if you merge properties into a schema, there is no way declare the properties as `required`... because you'd have to add the properties to the parent's `required` field, and the parent in our case is `schema.json, that's why we have to programmatically derive the list of `required` properties. I wish `ajv-merge-patch` can handle this, but unfortunately the library doesn't.

Reply:
so the solution i came up is to include `required` in the `customSchemas` optional argument, its format look like:


```js
[
    {
        key: 'commerceApi',
        required: true,
        schema: {/* actual schema */}
    }
]
```

Reply:
The downside of working around the limitation like this, is that there is this new format people need to understand... there is a tradeoff here, maybe we accept the fact that you can't declare you custom property as required. I'd like to get your thoughts on this.